src/models/sourcecodemodel.cpp

```diff
@@ -0,0 +1,148 @@
+/*
+    SPDX-FileCopyrightText: Lieven Hey <lieven.hey@kdab.com>
+    SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+
+    SPDX-License-Identifier: GPL-2.0-or-later
+*/
+
+#include "sourcecodemodel.h"
+#include <QDir>
+#include <QFile>
+
+SourceCodeModel::SourceCodeModel(QObject* parent)
+    : QAbstractTableModel(parent)
+{
+}
+
+SourceCodeModel::~SourceCodeModel() = default;
+
+void SourceCodeModel::clear()
+{
+    beginResetModel();
+    m_sourceCode.clear();
+    endResetModel();
+}
+
+void SourceCodeModel::setDisassembly(const DisassemblyOutput& disassemblyOutput)
+{
+    if (disassemblyOutput.sourceFileName.isEmpty())
+        return;
+
+    QFile file(m_sysroot + QDir::separator() + disassemblyOutput.sourceFileName);
+
+    if (!file.open(QIODevice::ReadOnly | QIODevice::Text))
+        return;
+
+    beginResetModel();
+
+    const auto lines = QString::fromUtf8(file.readAll()).split(QLatin1Char('\n'));
+
+    int maxIndex = 0;
+    int minIndex = INT_MAX;
+
+    m_validLineNumbers.clear();
+
+    for (const auto& line : disassemblyOutput.disassemblyLines) {
+        if (line.sourceCodeLine == 0)
+            continue;
+
+        if (line.sourceCodeLine > maxIndex) {
+            maxIndex = line.sourceCodeLine;
+        }
+        if (line.sourceCodeLine < minIndex) {
+            minIndex = line.sourceCodeLine;
+        }
+
+        m_validLineNumbers.insert(line.sourceCodeLine);
+    }
+
+    m_sourceCode.clear();
+
+    // the start / end of a function is defined by { }
+    // the following code tries to find ( to include the function signature in the source code
+    if (!lines[minIndex - 1].contains(QLatin1Char('('))) {
+        for (int i = minIndex - 1; i > minIndex - 6; i--) {
```

Comment:
merge the i < 0 check in here, i.e.:

```
for (int i = minIndex - 1; i >= 0 && i > minIndex - 6; --i) {
```

but what does the `-6` do? note how `minIndex--` is also changed in the body below (which is probably wrong?)

Reply:
the -6 means, that hotspot will search up to 6 lines for a `(` to find the function declaration.

In hindsight this is stupid. It would be much better to check if the line before minIndex is empty, of not include it until there is an empty line.
The reason is, DWARF starts a function at the `{` and not at the function declaration, but I would like to include the declaration. The code I wrote is just working for c++ which is kind of stupid.
The new variant should be a bit more robust, since most developers add an empty line between functions to split up the file visually